matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/relation/DefaultRelationService.kt

```diff
@@ -77,66 +75,72 @@ internal class DefaultRelationService @AssistedInject constructor(
                         .none { it.addedByMe && it.key == reaction }) {
             val event = eventFactory.createReactionEvent(roomId, targetEventId, reaction)
                     .also { saveLocalEcho(it) }
-            return eventSenderProcessor.postEvent(event, false /* reaction are not encrypted*/)
+            // TODO: This method should suspend!
+            eventSenderProcessor.postEvent(event, false /* reaction are not encrypted*/)
         } else {
             Timber.w("Reaction already added")
-            NoOpCancellable
         }
     }
 
-    override fun undoReaction(targetEventId: String, reaction: String): Cancelable {
+    override suspend fun undoReaction(targetEventId: String, reaction: String) {
         val params = FindReactionEventForUndoTask.Params(
                 roomId,
                 targetEventId,
                 reaction
         )
-        // TODO We should avoid using MatrixCallback internally
-        val callback = object : MatrixCallback<FindReactionEventForUndoTask.Result> {
-            override fun onSuccess(data: FindReactionEventForUndoTask.Result) {
-                if (data.redactEventId == null) {
-                    Timber.w("Cannot find reaction to undo (not yet synced?)")
-                    // TODO?
-                }
-                data.redactEventId?.let { toRedact ->
-                    val redactEvent = eventFactory.createRedactEvent(roomId, toRedact, null)
-                            .also { saveLocalEcho(it) }
-                    eventSenderProcessor.postRedaction(redactEvent, null)
-                }
+        val data = try {
+            // TODO We should avoid using MatrixCallback internally
+            awaitCallback<FindReactionEventForUndoTask.Result> {
+                findReactionEventForUndoTask
+                        .configureWith(params) {
+                            this.retryCount = Int.MAX_VALUE
+                            this.callback = it
+                        }
+                        .executeBy(taskExecutor)
```

Comment:
We could also call `execute()` (with no parameter now), and remove teh usage of the MatrixCallback, WDYT?
I can do it if necessary.